### ISSUE TITLE
Support Fira Code-style progress bars?

### ISSUE THREAD

**Issue by lovelymono** [MonoLisaFont/feedback#133]
Fira Code since 6.0 uses Unicode Private Use Area to provide dedicated glyphs for progress bars and loading animations. It would be really nice to have support for them in MonoLisa as well.

![image](https://user-images.githubusercontent.com/99364286/157404297-8b12bdae-fe50-4df0-90ff-4c3890cf2e53.png)

![image](https://i.imgur.com/q0Ms8n0.gif)

**Comment by okonet** [comment]
Thanks for the suggestion! It's a free at one. @MarcusSterz Do you want to take a look at this one?

**Comment by MarcusSterz** [comment]
Can do! Which unicode slots shall we use for this?

**Comment by lovelymono** [comment]
Preferably the same ones as Fira Code. It's EE00 to EE05 for progress bar, and EE06 to EE0B for spinner animation.

**Comment by MarcusSterz** [comment]
> Preferably the same ones as Fira Code. It's `EE00` to `EE05` for progress bar, and `EE06` to `EE0B` for spinner animation. (BTW, the first image has the Unicode codes for each glyph as well.)

I am on it. It there an explanation why the circle segments do not have the same shape but differ in size?

**Comment by lovelymono** [comment]
> Is there an explanation why the circle segments do not have the same shape but differ in size?

Not that I know of, no.
Asked for feedback at tonsky/FiraCode#1392.

**Comment by MarcusSterz** [comment]
@okonet @bebraw How is the status here? The design is ready.

**Comment by MarcusSterz** [comment]
> @MarcusSterz What do you mean by status?

The design is finished. Andrey wanted to create some animation to check if it looks good.

**Comment by bebraw** [comment]
> The design is finished. Andrey wanted to create some animation to check if it looks good.

Ah, ok.

Are the glyphs in the dev version? I guess he would need a font to test with.

@lovelymono Do you know specific terminal applications that use glyphs already?

**Comment by MarcusSterz** [comment]
Yes they are. 
<img width="1269" alt="Bildschirmfoto 2022-05-16 um 10 47 48" src="https://user-images.githubusercontent.com/51958720/168554915-07427c36-4613-4cf6-b575-ec4999d197c2.png">

**Comment by MarcusSterz** [comment]
> Are the glyphs in the dev version? I guess he would need a font to test with.
You should be able to acces them now.

**Comment by krish-r** [comment]
Hello,

I tried to play with the progress bar glyphs based on `v2.0.0 Prerelease` release notes.
However, I don't see them in `v2.0.0`. Can you please check?

|||
|-|-|
| Font Version | v2.0.0 Prerelease |
| Font format | variable ttf (Professional) |

![image](https://user-images.githubusercontent.com/54745129/187848594-019d9571-7718-4800-9af3-b57bada1c264.png)

```html
<!-- test.html -->

<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
    </head>
    <body style="padding: 25px;">
        <div style="font-family: 'MonoLisa'">
            <p>MonoLisa</p>
            <p>&#xEE00 &#xEE01 &#xEE02 &#xEE03 &#xEE04 &#xEE05</p>
            <p>&#xEE06 &#xEE07 &#xEE08 &#xEE09 &#xEE0A &#xEE0B</p>
        </div>
        <br>
        <div style="font-family: 'Fira Code'">
            <p>Fira Code</p>
            <p>&#xEE00 &#xEE01 &#xEE02 &#xEE03 &#xEE04 &#xEE05</p>
            <p>&#xEE06 &#xEE07 &#xEE08 &#xEE09 &#xEE0A &#xEE0B</p>
        </div>
    </body>
</html>
```

**Comment by bebraw** [comment]
@MarcusSterz Can you validate that the progress bar glyphs are included in the 2.0 build output? What's a good way to check which glyphs are included to a font on MacOS?

**Comment by bebraw** [comment]
These are from Font book.

**Fira Code:**

<img width="1041" alt="Screen Shot 2022-09-01 at 10 34 43" src="https://user-images.githubusercontent.com/166921/187858472-96d26d65-7d38-4bf4-bba9-91911fc104a6.png">

**MonoLisa:**

<img width="1021" alt="Screen Shot 2022-09-01 at 10 35 03" src="https://user-images.githubusercontent.com/166921/187858465-90bca405-9015-4438-8663-94412af45809.png">

It seems the glyphs are included but for a reason I don't understand the glyphs in MonoLisa don't have a UTF code associated to them but Fira Code does. This can be seen easily by hovering on a glyph. I think this is probably the reason why it doesn't render in browser as expected (I could confirm the behavior) but I don't understand fonts enough to tell for sure.

**Comment by MarcusSterz** [comment]
My bad! They are available but the unicodes were missing for some reason. Fixed it already and will be available in the next version.

**Comment by bebraw** [comment]
@krish-r There's a new prerelease (2.001) that should now contain the glyphs.

**Comment by krish-r** [comment]
wow! looks awesome! :clap:

especially -
* the use of rounded corners in the progress bar
* that there's more than variation of throbber/loading icon to choose from

![Peek 2022-09-23 14-16](https://user-images.githubusercontent.com/54745129/191924370-45cea1ce-875d-4ca1-871c-c1022f2888fc.gif)

![Peek 2022-09-23 14-43](https://user-images.githubusercontent.com/54745129/191929208-226e439b-dcd8-47a3-8f27-7acba386964a.gif)

<details>
<summary>Here's the quick and dirty bash script (if anyone's interested)</summary>

```bash
#!/usr/bin/env bash

# colors
c_green='\033[0;32m'
c_clear='\033[0m'

## progress bar
zero='\uEE00\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
ten='\uEE03\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
twenty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
thirty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
forty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
fifty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
sixty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
seventy='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
eighty='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE01\uEE02'
ninety='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE01\uEE01\uEE01\uEE02'
hundred='\uEE03\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE04\uEE05'
progress_bar=("$zero" "$ten" "$twenty" "$thirty" "$forty" "$fifty" "$sixty" "$seventy" "$eighty" "$ninety" "$hundred")
percentages=("0%" "10%" "20%" "30%" "40%" "50%" "60%" "70%" "80%" "90%" "100%")

printf "\n\n"
for ((i = 0; i < ${#progress_bar[@]}; i++)); do
	printf "\r     %b%b%b (%4s)" "$c_green" "${progress_bar[$i]}" "$c_clear" "${percentages[$i]}"
	sleep 0.50
done
printf "\r%100s\n\n" ""

## loading icon
loading_one='\uEE06'
loading_two='\uEE07'
loading_three='\uEE08'
loading_four='\uEE09'
loading_five='\uEE0A'
loading_six='\uEE0B'
loading=("$loading_one" "$loading_two" "$loading_three" "$loading_four" "$loading_five" "$loading_six")

loading_alt_one='\uEE10'
loading_alt_two='\uEE11'
loading_alt_three='\uEE12'
loading_alt_four='\uEE13'
loading_alt_five='\uEE14'
loading_alt_six='\uEE15'
loading_alt=("$loading_alt_one" "$loading_alt_two" "$loading_alt_three" "$loading_alt_four" "$loading_alt_five" "$loading_alt_six")

printf "\n\n"
for ((i = 0; i < 3; i++)); do
	for ((j = 0; j < ${#loading[@]}; j++)); do
		printf "\r     %b%b     %b     %b" "$c_green" "${loading[$j]}" "${loading_alt[$j]}" "$c_clear"
		sleep 0.25
	done
done
printf "\r%100s\n\n" ""

```

</details>

**Comment by MarcusSterz** [comment]
Cool – thanks for sharing!

**Comment by bebraw** [comment]
Published and included to 2.0.

**Comment by Bad3r** [comment]
@krish-r I refactored the script and put it in a repo:
https://github.com/Bad3r/progress-and-loading-bash/blob/main/progress_bar.sh

I hope that you don't mind. Thank you for sharing!
